tests/integration/Core/Framework/DataAbstractionLayer/Field/EnumFieldTest.php

```diff
@@ -0,0 +1,121 @@
+use Shopware\Core\Framework\Log\Package;
+use Shopware\Core\Framework\Test\TestCaseBase\KernelTestBehaviour;
+use Shopware\Core\Framework\Uuid\Uuid;
+use Shopware\Core\Framework\Validation\WriteConstraintViolationException;
+use Shopware\Tests\Integration\Core\Framework\DataAbstractionLayer\Field\EnumerationField\TestIntegerEnumeration;
+use Shopware\Tests\Integration\Core\Framework\DataAbstractionLayer\Field\EnumerationField\TestStringEnumeration;
+
+/**
+ * @internal
+ */
+#[Package('core')]
+#[Group('Field')]
+#[Group('DAL')]
+class EnumFieldTest extends TestCase
+{
+    use KernelTestBehaviour;
+
+    /**
+     * @param list<Flag> $flags
+     */
+    #[DataProvider('enumerationFieldDataProvider')]
+    public function testEnumerationFieldSerializer(
+        string $type,
+        mixed $input,
+        string|int|null $expected,
+        \BackedEnum $enum,
+        array $flags = []
+    ): void {
+        $serializer = static::getContainer()->get(EnumFieldSerializer::class);
+
+        $name = 'enum_' . Uuid::randomHex();
+        $data = new KeyValuePair($name, $input, false);
+
+        if ($type === 'writeException') {
+            $this->expectException(WriteConstraintViolationException::class);
+
+            try {
+                $serializer->encode(
+                    $this->getEnumerationField($name, $enum, $flags),
+                    EntityExistence::createEmpty(),
+                    $data,
+                    $this->getWriteParameterBagMock()
+                )->current();
+            } catch (WriteConstraintViolationException $e) {
+                static::assertSame('/' . $name, $e->getViolations()->get(0)->getPropertyPath());
+                static::assertSame($expected, $e->getViolations()->get(0)->getMessage());
+
+                throw $e;
+            }
+        }
+
+        if ($type === 'assertion') {
+            static::assertSame(
+                $expected,
+                $serializer->encode(
+                    $this->getEnumerationField($name, $enum, $flags),
+                    EntityExistence::createEmpty(),
+                    $data,
+                    $this->getWriteParameterBagMock()
+                )->current()
+            );
+        }
+    }
+
+    /**
+     * @return list<array{string, bool|string|int|\BackedEnum|\stdClass|null, string|\BackedEnum|null, \BackedEnum, array<Flag>}>
+     */
+    public static function enumerationFieldDataProvider(): array
+    {
+        $validationFailedForInt = 'This value should satisfy at least one of the following constraints: [1] This value should be of type integer. [2] This value should be null.';
+        $validationFailedForString = 'This value should satisfy at least one of the following constraints: [1] This value should be of type string. [2] This value should be null.';
+        return [
+            ['assertion', 'string', TestStringEnumeration::Regular->value, TestStringEnumeration::TrailingSpace, []],
```

Comment:
would be cool, if you add array keys as descriptions for each test case:

```suggestion
            'this and that causes foo and bar' => ['assertion', 'string', TestStringEnumeration::Regular->value, TestStringEnumeration::TrailingSpace, []],
```